### ISSUE TITLE
AudioStreamPlayer3D does not finish playing

### ISSUE THREAD

**Issue by cstaavetti** [godotengine/godot#54821]
### Godot version

Godot 4.0 dev snapshot (20211108, pre-alpha)  Built from commit 59686536622473e14b253268354ecfbae63d0d58

### System information

Arch Linux 5.14.16-arch1-1, Intel 3570K, Radeon 580

### Issue description

AudioStreamPlayer3D does not finish playing non looping audiostream.

After "play()" has been called, it will play the sound normally, but once it reaches the end "playing" state will continue to be "true".
"finished" signal seems to never be called and "get_playback_position()" will continue reporting the last position.

I was playing an .ogg file.
On 3.x the "finished" signal would be called and "playing" state would change to "false" at the end of the same file.

### Steps to reproduce

```gdscript
extends AudioStreamPlayer3D
@onready var audio_stream: AudioStream = preload("res://audio/explosion.ogg")

func _ready():
	set_stream(audio_stream)
	connect("finished", Callable(self, "end"))
	play()

func _physics_process(delta):
	print(playing, " ", get_playback_position())
	
func end():
	print("playing ended")
```

**Comment by Calinou** [comment]
@cstaavetti Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot. (We prefer ZIP archives or Git repositories to a code sample that needs to be copied into an existing project manually.)

**Comment by cstaavetti** [comment]
[cstaavetti_reproduction.zip](https://github.com/godotengine/godot/files/7510322/cstaavetti_reproduction.zip)

Here is a minimal reproduction and also a version for Godot 3 for comparison.

**Comment by Archie3d** [comment]
Seems like the `AudioStreamPlaybackResampled::mix` always returns the requested number of frames even when a non-looping stream gets depleted. Also there is this strange condition in that `mix()` function:
```cpp
if (idx <= internal_buffer_end && idx >= internal_buffer_end && mixed_frames_total == p_frames) {
	// The internal buffer ends somewhere in this range, and we haven't yet recorded the number of good frames we have.
	mixed_frames_total = i;
}
```

**Comment by ellenhp** [comment]
I'll have to go back and figure out what was going on there. Thanks for the bug report.

**Comment by cstaavetti** [comment]
I had some time on my hands and decided to take a deeper look at this issue.
The issue is actually Ogg Vorbis specific.

The issue seem to stem from audio_stream_ogg_vorbis.cpp
```cpp
int AudioStreamPlaybackOGGVorbis::_mix_frames_vorbis(AudioFrame *p_buffer, int p_frames) {
	...
	int frames = vorbis_synthesis_pcmout(&dsp_state, &pcm);
	if (frames > p_frames) {
		frames = p_frames;
		have_samples_left = true;
	} else {
		have_samples_left = false;
	}
	...
	return frames;
}
```

"vorbis_synthesis_pcmout" seem to always return either exactly "p_frames" or some seemingly random larger value that is then replaced by "p_frames".

This causes the "todo" to be subtracted down to 0 every time, which in turn causes "frames_mixed_this_step" to always be the same as "p_frames"
```cpp
int AudioStreamPlaybackOGGVorbis::_mix_internal(AudioFrame *p_buffer, int p_frames) {
	...
	int todo = p_frames;
	int start_buffer = 0;
	int frames_mixed_this_step = p_frames;
	while (todo && active) {
		...
		int mixed = _mix_frames_vorbis(buffer, todo);
		if (mixed < 0) {
			return 0;
		}
		todo -= mixed;
		...
		if (!have_packets_left) {
			//end of file!
			bool is_not_empty = mixed > 0 || vorbis_stream->get_length() > 0;
			if (vorbis_stream->loop && is_not_empty) {
				...
			} else {
				frames_mixed_this_step = p_frames - todo;
				for (int i = p_frames - todo; i < p_frames; i++) {
					p_buffer[i] = AudioFrame(0, 0);
				}
				active = false;
				todo = 0;
			}
		}
	}
	return frames_mixed_this_step;
}
```
After this "int AudioStreamPlaybackResampled::mix" in audio_stream.cpp fails to alter the value of "internal_buffer_end", is_playing() becomes true and the audio_stream_player is stuck.

It seems to me like this is either a bug with libvorbis or the return value of "vorbis_synthesis_pcmout" is handled incorrectly.
Or of course it could be something else, but anyways I am currently stuck here with fixing the issue.

**Comment by ellenhp** [comment]
Amazing work, thank you for figuring out what's happening. I think I know what's wrong. Fix incoming, just need a few hours.

**Comment by ellenhp** [comment]
@cstaavetti If you have a 4.0 development environment set up on your machine and have some spare time, would you mind testing #55808 to make sure it solves the issue on your end?

**Comment by cstaavetti** [comment]
Yeah, I'll definitely test the fix!

**Comment by cstaavetti** [comment]
I tested the fix #55808 and while it did indeed fix an issue there was still a second issue causing the same problem for some files.
I opened a pull request #55820 that should fix this and included a second reproduction of the issue with a working and broken ogg file.
